File: algorithm/leetcodeObjectiveC/sort-list.m
// 148. Sort List
// Difficulty: Medium

// Sort a linked list in O(n log n) time using constant space complexity.

// Time:  O(nlogn)
// Space: O(logn)

#import <Foundation/Foundation.h>

#pragma mark ListNode

@interface ListNode : NSObject

@property (atomic, assign) NSInteger value;
@property (atomic, strong) ListNode *next;

- (id)initWithValue:(NSInteger)value next:(ListNode *)next;
- (id)initWithValue:(NSInteger)value;
- (id)init;

@end

@implementation ListNode

- (id)initWithValue:(NSInteger)value next:(ListNode *)next {
  self = [super init];
  if (self) {
    self.value = value;
    self.next = next;
  }
  return self;
}

- (id)initWithValue:(NSInteger)value {
  return [self initWithValue:value next:nil];
}

- (id)init {
  return [self initWithValue:0 next:nil];
}

@end

#pragma mark Solution

ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
  ListNode* dummy = [[ListNode alloc] init];
  ListNode* curr = dummy;
  
  while (l1 && l2) {
    if (l1.value <= l2.value) {
      curr.next = l1;
      l1 = l1.next;
    } else {
      curr.next = l2;
      l2 = l2.next;
    }
    curr = curr.next;
  }
  curr.next = l1 ? l1 : l2;
  
  return dummy.next;
}

ListNode* sortList(ListNode* head) {
  if (!head || !head.next) {
    return head;
  }
  
  ListNode* slow = head, *fast = head;
  while (fast.next && fast.next.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  
  // Split linked list.
  ListNode* head2 = slow.next;
  slow.next = nil;
  
  return mergeTwoLists(sortList(head), sortList(head2));
}

File: data-structure-and-algorithm/javascript/leetcode/sort-list.js
/*
Given the head of a linked list, return the list after sorting it in ascending order.

Follow up: Can you sort the linked list in O(n logn) time and O(1) memory (i.e. constant space)?

 

Example 1:


Input: head = [4,2,1,3]
Output: [1,2,3,4]
Example 2:


Input: head = [-1,5,3,4,0]
Output: [-1,0,3,4,5]
Example 3:

Input: head = []
Output: []
 

Constraints:

The number of nodes in the list is in the range [0, 5 * 104].
-105 <= Node.val <= 105
*/

function sortList(head) {
    const dummy = new ListNode();
    dummy.next = head;

    let count = 0;
    for (; head; head = head.next, ++count);
    
    for (let i = 1; i < count; i *= 2) {
        for (let prev = dummy, curr = dummy.next; curr;) {
            const left = curr;
            const right = split(left, i);
            curr = split(right, i);
            prev = merge(left, right, prev);
        } 
    }
    
    return dummy.next;
}

function split(head, size) {
    if (!head) {
        return head
    }
    
    for (let i = 1; head.next && i < size; ++i) {
        head = head.next;
    }
    
    const right = head.next;
    head.next = null;
    return right;
}

function merge(left, right, prev) {
    for (; left && right; prev = prev.next) {
        if (left.val < right.val) {
            prev.next = left;
            left = left.next;
        } else {
            prev.next = right;
            right = right.next;
        }
    }
    
    prev.next = left ? left : right;
    for (; prev.next; prev = prev.next);
    return prev;
}

function sortList2(head) {
    if (!head || !head.next) {
        return head;
    }
    let prev = null, slow = head, fast = head;
    for (; fast && fast.next; slow = slow.next, fast = fast.next.next) {
        prev = slow;
    }

    if (fast) {
        prev = slow;
        slow = slow.next;
    }

    prev.next = null
    return merge(sortList(head), sortList(slow));

    function merge(l1, l2) {
        let dummy = new ListNode();
        let prev = dummy;
        for (; l1 && l2; prev = prev.next) {
            if (l1.val < l2.val) {
                prev.next = l1;
                l1 = l1.next;
            } else {
                prev.next = l2;
                l2 = l2.next;
            }
        }
        prev.next = l1 ? l1 : l2;
        return dummy.next;
    }
}